shanoir-ng-front/src/app/shared/mass-download/mass-download.service.ts

```diff
@@ -0,0 +1,492 @@
+
+declare var JSZip: any;
+
+export type Report = {
+    taskId: number,
+    folderName: string,
+    requestedDatasetIds: number[],
+    startTime: number,
+    list?: {
+        [key: number]: {
+            status: 'QUEUED' | 'ERROR' | 'SUCCESS',
+            error?: any,
+            errorTime?: number
+        }
+    }
+    nbSuccess?: number;
+    nbError?: number;
+    duration?: number;
+    format: Format;
+    nbQueues: number;
+};
+
+@Injectable()
+export class MassDownloadService {
+
+    private downloadQueue: Queue = new Queue();
+    readonly BROWSER_COMPAT_ERROR_MSG: string = 'browser not compatible';
+    readonly REPORT_FILENAME: string = 'downloadReport.json';
+
+    constructor(
+        private datasetService: DatasetService,
+        private notificationService: NotificationsService,
+        private consoleService: ConsoleService,
+        private dialogService: ConfirmDialogService) {
+    }
+
+    downloadByIds(datasetIds: number[]): Promise<void> {
+        return this.openModal().then(ret => {
+            if (ret != 'cancel') {
+                return this._downloadByIds(datasetIds, ret.format, ret.nbQueues)
+            } else return Promise.resolve();
+        }).catch(error => {
+            if (error == this.BROWSER_COMPAT_ERROR_MSG) {
+                return this.openAltModal().then(ret => {
+                    if (ret != 'cancel') {
+                        return this._downloadByIdsAlt(datasetIds, ret);
+                    } else return Promise.resolve();
+                });
+            } else throw error;
+        });
+    }
+
+    downloadDatasets(datasets: Dataset[]): Promise<void> {
+        return this.openModal().then(ret => {
+            if (ret != 'cancel') {
+                return this._downloadDatasets(datasets, ret.format, ret.nbQueues);
+            } else return Promise.resolve();
+        });
+    }
+
+    downloadAllByStudyId(studyId: number): Promise<void> {
+        return this.openModal().then(ret => {
+            if (ret != 'cancel') {
+                return this.datasetService.getByStudyId(studyId).then(datasets => {
+                    this._downloadDatasets(datasets, ret.format, ret.nbQueues);
+                })
+            } else return Promise.resolve();
+        });
+    }
+
+    downloadAllByStudyIdAndSubjectId(studyId: number, subjectId: number): Promise<void> {
+        return this.openModal().then(ret => {
+            if (ret != 'cancel') {
+                return this.datasetService.getByStudyIdAndSubjectId(studyId, subjectId).then(datasets => {
+                    return this._downloadDatasets(datasets, ret.format, ret.nbQueues).then(r => {
+                        return r;
+                    });
+                })
+            } else return Promise.resolve();
+        });
+    }
+
+    private _downloadByIds(datasetIds: number[], format: Format, nbQueues: number = 4, task?: Task, report?: Report, parentHandle?: FileSystemDirectoryHandle): Promise<void> {
```

Comment:
I wonder if it's a good idea to propose the user to have the choice in the number of queues. At least we should inform them what it means and what it changes (and maybe what is the most adapted for them ?)